### PR TITLE
Change stack-based regex state data to use global memory

### PR DESCRIPTION
**Pull request by davidwendt**
All libcudf strings regex calls will use global device memory for state data when evaluating regex on strings. Previously, separate templated kernels were used to store state data in fixed size stack memory depending on the number of instructions resolved from the provided regex pattern. This required the CUDA driver to allocate a large amount of device memory for when launching the kernel. This memory is managed by the launcher in the driver and so not under control of RMM.

This has been changed to use a memory-resource allocated global device memory to hold and manage the state data per string per instruction. This is an internal change only and results in no behavior changes. Overall, the performance based on the current benchmarks has not changed though much more memory may be required to execute any of the regex APIs depending on the number of instructions in the pattern and the total number of strings in the column.

Every effort has been made to not reduce performance from the stack-based approach. Additional optimizations here include copying the `reprog_device` class data to shared-memory (when it fits). Further optimizations are expected in later PRs as well.

Overall, the compile time of the files that use regex is also faster since only a single kernel is generated instead of 4 in the templated, stack-based implementation.

This PR is dependent on PR #10573.

### CODE REVIEW

**Comment by andygrove** [comment]
@davidwendt I ran some benchmarks and documented the results in https://github.com/NVIDIA/spark-rapids/issues/5178

The performance looks really good.

File: cpp/src/strings/regex/regexec.cu
Comment (hyperbolic2346) [inline cpp/src/strings/regex/regexec.cu:150]: Do you need to ever worry about the case where `buffer_size` is still > `max_size` after this while? If `thread_count` drops below `min_rows` you could leave the while with a `buffer_size` that is too large. I assume the converse can't be true as the buffer size required depends on the number of threads and if there are too many threads the buffer size will be large. Also, `thread_count` doesn't get forced back to `min_rows`. Is that variable named well or is it an issue that `thread_count` could be below `min_rows`?

File: cpp/src/strings/regex/utilities.cuh
Comment (hyperbolic2346) [inline cpp/src/strings/regex/utilities.cuh:36]: These names are very generic, but they are fairly specific in use. Is there a chance that another `strings::detail` kernel would be named similarly? Maybe `for_each_re_kernel`?

File: cpp/src/strings/regex/regexec.cu
Comment (davidwendt) [reply cpp/src/strings/regex/regexec.cu:150]: It is ok if the `buffer_size` is greater than `max_size` if we hit the `min_rows`. I see what you mean about the `thread_count` going below `min_rows`. The intention was the `min_rows` restriction would override the `max_size` one.

File: cpp/src/strings/regex/utilities.cuh
Comment (davidwendt) [reply cpp/src/strings/regex/utilities.cuh:36]: It is certainly possibly for the name to be reused but not with these parameters. I felt the `reprog_device` parameter specifically communicated this was to be used with regex kernels.

File: cpp/src/strings/regex/regexec.cu
Comment (davidwendt) [reply cpp/src/strings/regex/regexec.cu:155]: I've updated the logic to clamp to `min_rows` and update the doxygen to explain the behavior. Let me know what you think. The intention is to expose this through an API in a later PR to help callers predict memory usage for given regex pattern and column size.

Comment (hyperbolic2346) [reply cpp/src/strings/regex/regexec.cu:155]: This makes sense to me and I like the changes you made, thank you for that.

Is `max_size` a good name here if it isn't the max size. Maybe `preferred_size` or `desired_max`?

Comment (davidwendt) [reply cpp/src/strings/regex/regexec.cu:155]: I'll change it to `requested_max_size`.

Comment (hyperbolic2346) [reply cpp/src/strings/regex/regexec.cu:155]: Thanks for humoring my pedantic requests. 🥇

File: cpp/src/strings/extract/extract_all.cu
Comment (vyasr) [inline cpp/src/strings/extract/extract_all.cu:55]: Why is this an `int32_t` rather than a `size_type`? Other parts of this PR have moved towards greater consistency in using `size_type`.

File: cpp/src/strings/regex/utilities.cuh
Comment (vyasr) [reply cpp/src/strings/regex/utilities.cuh:98]: Should 256 just be `MINIMUM_THREADS`?

File: cpp/src/strings/extract/extract_all.cu
Comment (vyasr) [reply cpp/src/strings/extract/extract_all.cu:55]: On further reading, I guess you are using `int32_t` everywhere now for the number of threads or the number of distinct reprog instances while reserving `size_type` for its usual meaning of the max column size in libcudf?

File: cpp/src/strings/regex/regex.cuh
Comment (vyasr) [inline cpp/src/strings/regex/regex.cuh:151]:
```suggestion
    int32_t const rows,
    int32_t const min_rows               = MINIMUM_THREADS,
    std::size_t const requested_max_size = MAX_WORKING_MEM) const;
```

Comment (vyasr) [reply cpp/src/strings/regex/regex.cuh:151]: Probably can make a few other function parameters in this file constant, but I guess they're mostly host functions so it's not a huge deal.

**Review by vyasr — CHANGES_REQUESTED** [review]
This is just a first pass of review, I think I'll need one more to fully grasp the import of the changes to the regex internals. I have some probably naive questions that will probably make sense on a second read, but I'll ask them anyway in case they have non-obvious answers:
- Why did all the functors change from owning a `reprog_device` to having the operator accept one? Do you observe better memory access patterns (more L1 traffic due to the parameter being cached) by passing it as a `const` parameter this way?
- Am I correct that you are using `prog_idx` as the name for the third functor parameter when you `launch_transform_kernel` while using `thread_idx` as the name for `launch_foreach_kernel` because that more accurately reflects what each thread is acting on? I'm trying to see if there are helpful naming conventions that could help improve the clarity here.
- I'm not sure that I fully understand the usage of the new `load`/`store` methods and the utilization of shared memory. It looks like the initial state of the `reprog_device` is stored into shared memory from one thread, then it's loaded onto all the threads. Is the hope that in the cases where the object fits into shared memory the parameter passed in will be discarded (or at least relegated to a much lower position in the cache hierarchy) once the code starts using the lane 0 version copied into shared memory? Is there no way to make that determination prior to the kernel launch?

File: cpp/src/strings/contains.cu
Comment (davidwendt) [reply cpp/src/strings/contains.cu:58]: This is used by `contains_re` and `matches_re` so I did not want to name it just `contains`.

File: cpp/src/strings/regex/utilities.cuh
Comment (davidwendt) [reply cpp/src/strings/regex/utilities.cuh:98]: This is the one that seemed to work the best overall. It has only a loose relationship with `MINIMUM_THREADS` . I can give it a constexpr name.

File: cpp/src/strings/regex/utilities.cuh
Comment (davidwendt) [reply cpp/src/strings/regex/utilities.cuh:86]: @hyperbolic2346  already asked this question. Here was my response.
https://github.com/rapidsai/cudf/pull/10600#discussion_r859730787

File: cpp/src/strings/extract/extract_all.cu
Comment (davidwendt) [reply cpp/src/strings/extract/extract_all.cu:55]: Yes, it is not a `size_type` since the value is not really a size.

File: cpp/src/strings/regex/regex.cuh
Comment (davidwendt) [reply cpp/src/strings/regex/regex.cuh:151]: I don't think declaring primitive parameter types that are passed by value `const` makes much difference and does not communicate anything special to the caller.

**Comment by davidwendt** [comment]
> * Why did all the functors change from owning a `reprog_device` to having the operator accept one? Do you observe better memory access patterns (more L1 traffic due to the parameter being cached) by passing it as a `const` parameter this way?

Yes, we found this to helped with performance. Passing by value allows the object to live in register variables and reduced reads from global memory.

**Comment by davidwendt** [comment]
> * Am I correct that you are using `prog_idx` as the name for the third functor parameter when you `launch_transform_kernel` while using `thread_idx` as the name for `launch_foreach_kernel` because that more accurately reflects what each thread is acting on? I'm trying to see if there are helpful naming conventions that could help improve the clarity here.

Yes, that sounds right. The `reprog_device` is sort of ignorant of threads. It just needs an index to locate where to put its state data in the global state memory buffer.

**Comment by davidwendt** [comment]
> * I'm not sure that I fully understand the usage of the new `load`/`store` methods and the utilization of shared memory. It looks like the initial state of the `reprog_device` is stored into shared memory from one thread, then it's loaded onto all the threads. Is the hope that in the cases where the object fits into shared memory the parameter passed in will be discarded (or at least relegated to a much lower position in the cache hierarchy) once the code starts using the lane 0 version copied into shared memory? Is there no way to make that determination prior to the kernel launch?

I'm not sure I'm following your question here. The `reprog_device` data is stored in global memory and must be copied to shared-memory. The usual pattern is for the first thread of the block to this work and sync. The extra step here is to create an appropriate stack variable that wraps this data for each of the block threads. 
If the object data does not fit in shared memory, no shared memory is created at launch and the load/store functions do nothing.

File: cpp/src/strings/contains.cu
Comment (vyasr) [reply cpp/src/strings/contains.cu:58]: That makes sense, perhaps a different common regex term like `find` or `search`? Or maybe just a different suffix `contains_internal`? I just don't think `util` in particular is appropriate here, even `impl` would be a more accurate description of what this function is.

File: cpp/src/strings/regex/utilities.cuh
Comment (vyasr) [reply cpp/src/strings/regex/utilities.cuh:38]: I suspect that you're right and that while this is coming up during review (for both Mike and me) it will probably be clear to readers in the future.

File: cpp/src/strings/regex/regex.cuh
Comment (vyasr) [reply cpp/src/strings/regex/regex.cuh:151]: It can make a difference in device code where inlined functions can reuse the same values in registers from previous functions if they're all const, which in turn reduces memory traffic and register pressure, but I realized after I wrote this suggestion that this is host code and not performance-critical which is why I discounted it in my second comment. We can resolve this.

File: cpp/src/strings/regex/utilities.cuh
Comment (davidwendt) [reply cpp/src/strings/regex/utilities.cuh:38]: I'll have to disagree with this one. Will this prevent approval of this PR?

**Comment by davidwendt** [comment]
> 1. Whether or not `d_prog` will fit in shared memory is a runtime constant, not something known at compile time. 

The `d_prog` object is generated base on the given regex pattern so its size is only known at runtime. 

>This optimization relies on the runtime recognizing that `d_prog` is not accessed on any thread > 0 and therefore not accessing it in global memory, right? 

I'm not following this question. The `d_prog` is accessed by all threads (after the syncthreads). The first thread in the block loads the `d_prog` data into shared-memory. This is a common usage pattern for loading shared-memory.

>Otherwise you'll still initially load `d_prog` from global memory on all threads. Is that something we can rely on the runtime doing correctly, or is the idea that even with that initial load of `d_prog` from global memory it's worthwhile to access shared memory in all subsequent accesses to `s_prog` because the original `d_prog` will immediately be evicted from the cache if it goes unused?

The `d_prog` will either fit into shared-memory or not. If it fits, the `store` copies the `d_prog` into the shared-memory buffer and the `load` will give a new `d_prog` using the shared-memory. If it does not fit, the `store` does nothing and the `load` will just return whatever is passed to it.  Either way the `d_prog` returned from `load` will be valid for every thread to use.

**Comment by vyasr** [comment]
@davidwendt and I sorted most of this out offline. tl;dr the global memory buffers that `d_prog` points to are much larger than the object itself, which is essentially a handful of integers and pointers. Most of my suggestions boiled down to asking whether we could avoid `d_prog` being copied/used on all threads, but since the object is so small it's not worthwhile to try to optimize that out. The shared memory buffer is already doing the important work. 

I'm pretty much happy with the current state of the PR, but plan to give it another review with a fresh eye on Monday before approving.

File: cpp/src/strings/contains.cu
Comment (vyasr) [reply cpp/src/strings/contains.cu:58]: Great meme :1st_place_medal: Sadly, a great meme does not help to convince me that the naming here makes sense. 

Right now, `contains_util` is to `contains` basically what our detail APIs are to our public APIs, except with the `beginning_only` parameter instead of the `stream`. `contains_re` and `matches_re` delegate all their logic to this internal function. `util` implies a much smaller scope than that to me.

I don't want to block the PR over this, though, so I'll let @hyperbolic2346 weigh in with his thoughts. If he doesn't have an issue here then we can move forward, otherwise we can try to brainstorm a happy medium together.

**Review by vyasr — APPROVED** [review]
This LGTM now. I would still be interested to see whether these kernels are ever occupancy bound. If they are, then making sure that every thread uses a pointer to the `reprog_device` in shared memory rather than having to copy all of its members into a thread-local object might help you see even more performance gains by reducing register usage. That change can be explored in a future PR though.

File: cpp/src/strings/contains.cu
Comment (hyperbolic2346) [reply cpp/src/strings/contains.cu]: I'm not so strongly opinionated that I would build a meme for my argument. I'm fine with leaving this as-is.

**Review by hyperbolic2346 — APPROVED** [review]
Thanks for @vyasr for adding a comment explaining the offline discussion. This looks good to me and is a nice win.

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**
